go/mysqlconn/replication/binlog_event_json.go

```diff
@@ -0,0 +1,477 @@
+	val := uint32(data[0]) +
+		uint32(data[1])<<8 +
+		uint32(data[2])<<16 +
+		uint32(data[3])<<24
+	if toplevel {
+		result.WriteByte('\'')
+	}
+	result.Write(strconv.AppendInt(nil, int64(int32(val)), 10))
+	if toplevel {
+		result.WriteByte('\'')
+	}
+}
+
+func printJSONUint32(data []byte, toplevel bool, result *bytes.Buffer) {
+	val := uint32(data[0]) +
+		uint32(data[1])<<8 +
+		uint32(data[2])<<16 +
+		uint32(data[3])<<24
+	if toplevel {
+		result.WriteByte('\'')
+	}
+	result.Write(strconv.AppendUint(nil, uint64(val), 10))
+	if toplevel {
+		result.WriteByte('\'')
+	}
+}
+
+func printJSONInt64(data []byte, toplevel bool, result *bytes.Buffer) {
+	val := uint64(data[0]) +
+		uint64(data[1])<<8 +
+		uint64(data[2])<<16 +
+		uint64(data[3])<<24 +
+		uint64(data[4])<<32 +
+		uint64(data[5])<<40 +
+		uint64(data[6])<<48 +
+		uint64(data[7])<<56
+	if toplevel {
+		result.WriteByte('\'')
+	}
+	result.Write(strconv.AppendInt(nil, int64(val), 10))
+	if toplevel {
+		result.WriteByte('\'')
+	}
+}
+
+func printJSONUint64(data []byte, toplevel bool, result *bytes.Buffer) {
+	val := binary.LittleEndian.Uint64(data[:8])
+	if toplevel {
+		result.WriteByte('\'')
+	}
+	result.Write(strconv.AppendUint(nil, val, 10))
+	if toplevel {
+		result.WriteByte('\'')
+	}
+}
+
+func printJSONDouble(data []byte, toplevel bool, result *bytes.Buffer) {
+	val := binary.LittleEndian.Uint64(data[:8])
+	fval := math.Float64frombits(val)
+	if toplevel {
+		result.WriteByte('\'')
+	}
+	result.Write(strconv.AppendFloat(nil, fval, 'E', -1, 64))
+	if toplevel {
+		result.WriteByte('\'')
+	}
+}
+
+func printJSONString(data []byte, toplevel bool, result *bytes.Buffer) {
+	size, pos := readVariableInt(data, 0)
+
+	// A toplevel JSON string is printed as a JSON-escaped
+	// string inside a string, as the value is parsed as JSON.
+	// So the value should be: '"value"'.
+	if toplevel {
+		result.WriteString("'\"")
+		// FIXME(alainjobart): escape reserved characters
+		result.Write(data[pos : pos+size])
+		result.WriteString("\"'")
+		return
+	}
+
+	// Inside a JSON_ARRAY() or JSON_OBJECT method, we just print the string
+	// as SQL string.
+	valStr := sqltypes.MakeTrusted(sqltypes.VarBinary, data[pos:pos+size])
+	valStr.EncodeSQL(result)
+}
+
+func printJSONOpaque(data []byte, toplevel bool, result *bytes.Buffer) error {
+	typ := data[0]
+	size, pos := readVariableInt(data, 1)
+
+	// A few types have special encoding.
+	switch typ {
+	case TypeDate:
+		return printJSONDate(data[pos:pos+size], toplevel, result)
+	case TypeTime:
+		return printJSONTime(data[pos:pos+size], toplevel, result)
+	case TypeDateTime:
+		return printJSONDateTime(data[pos:pos+size], toplevel, result)
+	case TypeNewDecimal:
+		return printJSONDecimal(data[pos:pos+size], toplevel, result)
+	}
+
+	// FIXME(alainjobart) this is broken for now. The lack of metadata
+	// makes the parsing fail. The MySQL source code is too obscure
+	// to make any sense of this.
```

Comment:
Should we just return an error here?
If we know what json construct could cause this, we can also document it, and may be block it from vttablet.